REF/PERF: move np.errstate out of core array_ops up to higher level

jorisvandenbossche opened the pull request:
Currently, we suppress numpy warnings with `np.errstate(all="ignore")` inside the `evaluate` function in `core.computation.expressions`, which is is only used in `core.ops.array_ops._na_arithmetic_op`, which in itself is only used in `core.ops.array_ops` `arithmetic_op()` and `comparison_op()` (where we actually called `np.errstate` again, duplicatively). 

So, in summary, we suppress the warnings at the level of the "array op". For the ArrayManager, we call this array op many times for each column, and repeatedly calling `np.errstate(all="ignore")` gives a big overhead. Luckily, it is easy to suppress the warnings once at a higher level, at the DataFrame/Series level, where those array ops are called.

That's what this PR is doing: removing `np.errstate(all="ignore")` in the actual array ops, and adding it in all places where we currently call the array ops. 

With the benchmark case of an arithmetic op with two dataframes, this gives a considerable improvement:

```python
import numpy as np
import pandas as pd

df = pd.DataFrame(np.random.randn(1000, 1000))
df_am = df._as_manager("array")
```

```
In [2]: %timeit df_am + df_am
18.1 ms ± 1.04 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)   <-- master
8.57 ms ± 167 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  <-- PR
```

For BM it doesn't matter for this case, since that's a single block and `np.errstate` would be called only once anyway. But for certain cases of `df+s` ops where we potentially also work column-wise for BM, it should benefit there as well.

jorisvandenbossche committed:
REF/PERF: move np.errstate out of core array_ops up to higher level

jbrockmendel commented on pandas/core/ops/__init__.py:
```diff
@@ -433,7 +433,8 @@ def f(self, other, axis=default_axis, level=None, fill_value=None):
             # Another DataFrame
-            new_data = self._combine_frame(other, na_op, fill_value)
+            with np.errstate(all="ignore"):
+                new_data = self._combine_frame(other, na_op, fill_value)
```
shouldnt we get this one for free bc it goes through _dispatch_frame_op which you've got setting np.errstaet?

jorisvandenbossche replied:
Ah, yes, indeed `_combine_frame` eventually always goes through `_dispatch_frame` as well (I got here because it's a place that uses `get_array_op`)

jbrockmendel commented on pandas/core/ops/array_ops.py:
```diff
@@ -265,8 +268,7 @@ def comparison_op(left: ArrayLike, right: Any, op) -> ArrayLike:
         with warnings.catch_warnings():
             # suppress warnings from numpy about element-wise comparison
             warnings.simplefilter("ignore", DeprecationWarning)
-            with np.errstate(all="ignore"):
```
should the catch_warnings here be done at a higher level too?

jorisvandenbossche replied:
Possibly, but that might need to be done more targeted for the specific type of op (since we don't do anything similar in arithmetic_op), and thus this will be more complicated. So would like to start with `np.errstate`.

jbrockmendel commented:
This is unfortunate, since it is both a) uglier to do this at a higher level and b) a really nice perf improvement

jorisvandenbossche commented:
Is it that much uglier? I removed it in 5 places, and I added it in 4 places in `_dispatch_frame_op` (which in theory could be combined into a single one) and 3 other places. So it's not that big of a difference.

jorisvandenbossche committed:
remove in _combine_frame

jorisvandenbossche commented on pandas/core/arrays/numpy_.py:
```diff
@@ -376,7 +376,8 @@ def _cmp_method(self, other, op):
         pd_op = ops.get_array_op(op)
-        result = pd_op(self._ndarray, other)
+        with np.errstate(all="ignore"):
+            result = pd_op(self._ndarray, other)
```
BTW, we probably don't need to add this one here, since when going through a DataFrame/Series op, this will also already be catched on that level (and be don't consistently catch it in other EA ops). But this keeps current behaviour (and this not used in practice anyway)

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -6556,7 +6556,8 @@ def _dispatch_frame_op(self, right, func, axis: Optional[int] = None):
         if not is_list_like(right):
             # i.e. scalar, faster than checking np.ndim(right) == 0
-            bm = self._mgr.apply(array_op, right=right)
+            with np.errstate(all="ignore"):
```
could just do the np.errstate once at the top of the method?

jorisvandenbossche replied:
Yes, that is possible (as mentioned in https://github.com/pandas-dev/pandas/pull/40396#issuecomment-797660478). Personally, I think I prefer putting it just around the mgr operate call / array_op calls, but in the end it doesn't matter much (it's not that the other checks here could potentially gives such warning that would be incorrectly suppressed). So either way.

jbrockmendel replied:
im fine with this, we can clean it up later if needbe

jbrockmendel commented on pandas/core/series.py:
```diff
@@ -5087,7 +5088,8 @@ def _arith_method(self, other, op):
         rvalues = extract_array(other, extract_numpy=True)
-        result = ops.arithmetic_op(lvalues, rvalues, op)
+        with np.errstate(all="ignore"):
+            result = ops.arithmetic_op(lvalues, rvalues, op)
```
do we not need this for _logical_method?

jorisvandenbossche replied:
Currently we don't suppress any warnings in `ops.logical_op`, so I left the current behaviour as is.

jbrockmendel replied:
makes sense

jreback reviewed (approved):
looks reasonable

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into ops-refactor-seterr

jbrockmendel commented:
LGTM, merging this afternoon unless someone chimes in to stop me

jreback commented:
thanks @jorisvandenbossche
